### PR TITLE
store/copr: optimize copIterator by avoid start new goroutine

### PR DESCRIPTION
**Pull request by crazycs520**
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: ref #56649

Problem Summary: optimize cop iterator by avoid start new goroutine

### What changed and how does it work?

Before this PR, `copIterator` always start at least 1 [copIteratorWorker goroutine](https://github.com/pingcap/tidb/blob/f347b9d334495c72d27c53532dfef522213dbaac/pkg/store/copr/coprocessor.go#L884) and 1 [copIteratorTaskSender goroutine](https://github.com/pingcap/tidb/blob/f347b9d334495c72d27c53532dfef522213dbaac/pkg/store/copr/coprocessor.go#L902). That's too wasteful for small queries.

This PR avoid start new goroutine for small queries, for better performance, following is the sysbench [oltp_index_scan](https://github.com/crazycs520/sysbench_lua) test, and the range size is 2:

```sh
sysbench --config-file=sysbench.conf oltp_index_scan --tables=16 --table-size=1000000 --threads=100 --range-size=2 run
```

The example query is like this:

```sql
SELECT
  k
FROM
  sbtest1
WHERE
  k BETWEEN ?
  AND ? [arguments: (503925, 503926)]
```

Which Plan is:

```plan

| id                 | estRows | estCost | actRows | task      | access object               | execution info                                                                                                                                                                                                                                                                                         | operator info                           | memory    | disk  |
| IndexReader_6      | 2.07    | 31.20   | 2       | root      |                             | time:1.31ms, loops:2, cop_task: {num: 1, max: 1.15ms, proc_keys: 2, tot_proc: 81.2µs, tot_wait: 575.8µs, copr_cache: disabled, build_task_duration: 12.2µs, max_distsql_concurrency: 1}, rpc_info:{Cop:{num_rpc:1, total_time:1.12ms}}                                                                 | index:IndexRangeScan_5                  | 278 Bytes | N/A   |
| └─IndexRangeScan_5 | 2.07    | 336.86  | 2       | cop[tikv] | table:sbtest1, index:k_1(k) | tikv_task:{time:0s, loops:1}, scan_detail: {total_process_keys: 2, total_process_keys_size: 92, total_keys: 3, get_snapshot_time: 539.6µs, rocksdb: {key_skipped_count: 2, block: {cache_hit_count: 6}}}, time_detail: {total_process_time: 81.2µs, total_wait_time: 575.8µs, tikv_wall_time: 751.3µs} | range:[503925,503926], keep order:false | N/A       | N/A   |
```

version | workload | thread | QPS | QPS Increase
-- | -- | -- | -- | --
master | oltp_index_scan | 10 | 9224 |  
This PR | oltp_index_scan | 10 | 9471 | 2.6%
master | oltp_index_scan | 50 | 36072 |  
This PR | oltp_index_scan | 50 | 37392 | 3.6%
master | oltp_index_scan | 100 | 41977 |  
This PR | oltp_index_scan | 100 | 43554 | 3.7%
master | oltp_index_scan | 150 | 42785 |  
This PR | oltp_index_scan | 150 | 46430 | 8.5%



### Check List

Tests <!-- At least one of them must be included. -->

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/store/copr/coprocessor.go
Comment (zyguan) [inline pkg/store/copr/coprocessor.go:1174]: Is it possible that there is a remaining task whose batchTaskList is not empty (then multiple copResponses might be pushed into the respCh)?

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [inline pkg/store/copr/coprocessor.go:1152]: The original `ctx` could be used for the logger..

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [inline pkg/store/copr/coprocessor.go:1158]: Better to check the presumed constraints for the `it.tasks` before the `for` loop.

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [inline pkg/store/copr/coprocessor.go:1926]: Why not checking the `worker.kvclient.Stats` first so the map creations could be avoided in some cases?

File: pkg/store/copr/coprocessor.go
Comment (crazycs520) [reply pkg/store/copr/coprocessor.go:1174]: Nice catch, but the `it.liteWorker` only work when there is only 1 task and doesn't have `batchTaskList`, so this case won't happen. https://github.com/pingcap/tidb/pull/57522/files#diff-f341da742ab4e8413960a536dfe0038a2bb92ba2dd6e9f336b5708e19a8b3e60R861

File: pkg/store/copr/coprocessor.go
Comment (crazycs520) [reply pkg/store/copr/coprocessor.go:1158]: https://github.com/pingcap/tidb/pull/57522/files#diff-f341da742ab4e8413960a536dfe0038a2bb92ba2dd6e9f336b5708e19a8b3e60R861 Already checked `it.tasks`.

File: pkg/store/copr/coprocessor.go
Comment (zyguan) [reply pkg/store/copr/coprocessor.go:1174]: But what if the first task meets region error and newly tasks do have `batchTaskList`?

File: pkg/store/copr/coprocessor.go
Comment (crazycs520) [reply pkg/store/copr/coprocessor.go:1926]: It has been checked outside before call this function.

File: pkg/store/copr/coprocessor.go
Comment (crazycs520) [reply pkg/store/copr/coprocessor.go]: great, done.

File: pkg/store/copr/coprocessor.go
Comment (crazycs520) [reply pkg/store/copr/coprocessor.go:1093]: Nice catch, addressed, look like only one cop-reader can use cop-lite-worker.

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [inline pkg/store/copr/coprocessor.go:862]: This is not atomic operation, the `CAS` should used instead.

But why it's designed to be read-write concurrently? Should the `open` be a single-threaded operation?

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [reply pkg/store/copr/coprocessor.go:1159]: Sharing internal states across multiple functions and assuming that some states are manipulated in other functions is not a robust programming practice. 

Functions should minimize mutual state dependencies as much as possible, and dependencies should be passed explicitly through parameters.

Or at least, the cross-functions assumptions should be stated clearly in the comments.

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [reply pkg/store/copr/coprocessor.go:1930]: Ditto.

File: pkg/store/copr/coprocessor.go
Comment (crazycs520) [reply pkg/store/copr/coprocessor.go:862]: In some cases (such as index lookup reader), there are maybe some `TableReader` execute `open` concurrently.

File: pkg/store/copr/coprocessor.go
Comment (crazycs520) [reply pkg/store/copr/coprocessor.go:1187]: Nice catch, already fix it.

File: pkg/store/copr/coprocessor.go
Comment (crazycs520) [reply pkg/store/copr/coprocessor.go:1159]: After modifications, there are no assumptions for `it.tasks` here.

File: pkg/store/copr/coprocessor.go
Comment (crazycs520) [reply pkg/store/copr/coprocessor.go:1933]: ok, already add the check here.

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [inline pkg/store/copr/coprocessor.go:862]: How about using the `boolean` type for `tryCopLiteWorker` if its value is just on or off instead of checking `pointer == nil`?

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [reply pkg/store/copr/coprocessor.go:1093]: Better to add comments about the to be used scenarios above the `cop-lite-worker`.

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [inline pkg/store/copr/coprocessor.go:1172]: How about moving the check to L 1159 instead of doing it in the `for` loop?

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [reply pkg/store/copr/coprocessor.go:1187]: Is there a test case to cover this situation?

File: pkg/store/copr/coprocessor.go
Comment (zyguan) [inline pkg/store/copr/coprocessor.go:1175]: Here we assume that all copResponses are fill to the respChan after calling handleTaskOnce, how about just let handleTaskOnce return a slice of copResponses (instead of using a channel)?

```go
		for len(it.respChan) > 0 {
			select {
			case resp = <-it.respChan:
				return resp
			default:
			}
		}
```

The above code also assume that no copResponse will be pushed into the respChan asynchronously, but actually current code cannot guarantee that (eg. someone may change the impl of handleTaskOnce and let it write to respCh asynchronously).

File: pkg/store/copr/coprocessor.go
Comment (you06) [reply pkg/store/copr/coprocessor.go]: The workers under the same session is individual, I think the lite worker can be used in different `TableReader`s under one index lookup executor. However, the `DistSQLContext` is shared in session level.

By using `if atomic.CompareAndSwapUint32(tryCopLiteWorker, 0, 1) {...}`, only one cop iterator can benefit from the lite optimization.

File: pkg/store/copr/coprocessor.go
Comment (crazycs520) [reply pkg/store/copr/coprocessor.go:862]: I don't understand, CAS operations need to be done here, `boolean`  can't implement?

File: pkg/store/copr/coprocessor.go
Comment (crazycs520) [reply pkg/store/copr/coprocessor.go:1175]: Nice catch, I'm working on this.

File: pkg/distsql/context/context.go
Comment (you06) [inline pkg/distsql/context/context.go:90]: I'm inclined that the `TryCopLiteWorker` is a [request](https://github.com/pingcap/tidb/blob/a3d60bcfc6c1f18c67eeb319fc70cf26befb8687/pkg/kv/kv.go#L529-L530) level option.

Like an index lookup with 1 row result set. If `TryCopLiteWorker` is a session level option, it's only enabled once in index scan, and in table lookup, the `atomic.CompareAndSwapUint32(tryCopLiteWorker, 0, 1)` will fail and execute with multi-goroutine model.

Comment (crazycs520) [reply pkg/distsql/context/context.go:90]: I'm afraid that make `TryCopLiteWorker` in `request` level option may have similar problem in this [comment](https://github.com/pingcap/tidb/pull/57522#discussion_r1875515863), there is a known case and I have added [test](https://github.com/pingcap/tidb/pull/57522/commits/a3c7e530163c2572dc988dbe908b509edae45bc3#diff-68cc9873623c1d811b385e0f6786521e0a05ff90b96678508881406c6a4d5e8dR321) for it. And I'm not sure if there are other cases like this, so I prefer keep `TryCopLiteWorker ` to be statement level.

As index lookup with 1 row result set, you are right, but because of the above problems, I don't have a good way to fix this.

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [reply pkg/store/copr/coprocessor.go]: @crazycs520 
Is it possible to make the lite worker to be used for all the table workers?

File: pkg/store/copr/coprocessor.go
Comment (cfzjywxk) [reply pkg/store/copr/coprocessor.go]: @you06 @crazycs520 
Is it possible to trigger the kv request sending in the `Open` function call, so the result could be prepared doing Next?

**Review by zyguan — APPROVED** [review]

File: pkg/store/copr/coprocessor.go
Comment (you06) [reply pkg/store/copr/coprocessor.go]: > Is it possible to trigger the kv request sending in the `Open` function call

It's possible that the batch kv client returns a future, but it needs big changes to the kv client's interface.

**Review by you06 — COMMENTED** [review]
LGTM, once we have async kv interface, we can remove `TryCopLiteWorker` and make more coprocessor tasks benefit from this opimization.
Wait for @cfzjywxk's approve.

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**
